### PR TITLE
Some Tech Tree updates from Iain

### PR DESCRIPTION
**Pull request by mgaudet**
Iain did a brainstorming session; I embedded his output into the tech tree.

There's also a bit of clarification in this push.

### CODE REVIEW

File: tech_tree/diagram.mmd [hunk 1]
```diff
@@ -1,4 +1,4 @@
-%%{init: {"flowchart": {"defaultRenderer": "elk"}} }%%
+%%{init: {"flowchart": {"defaultRenderer": "default"}} }%%
 %% Use the elk renderer: Much better for this diagram.
```
Comment (iainireland) [inline tech_tree/diagram.mmd]: I don't have any opinion on which renderer to use, but if we're going to change it then maybe this comment should be updated?

File: tech_tree/index.md [hunk 2]
```diff
@@ -155,6 +155,25 @@ would be the creation of mid-tier optimizing compiler that's more of a divergenc
...
+## A Replacement DSL for Self-Hosted code {#selfHostedReplacement}
+
+Iain's wishlist:
+
+>  ergonomic to write, generates efficient code, can be inlined, multiple return values, a pony
+>
+> Most interesting syntax idea: `specialize (condition)` statement:
+> - generate two copies of the rest of the block
+>   - one specialized for the fast path
+> - one fallback path
```
Comment (iainireland) [inline tech_tree/index.md]: Nit: `one fallback path` should be aligned with `one specialized for the fast path`

File: tech_tree/diagram.mmd [hunk 3]
```diff
@@ -117,6 +145,34 @@ flowchart LR
...
+    click precompileBaselineICs href "https://bugzilla.mozilla.org/show_bug.cgi?id=1357884"
+
+    shareBaselineICs["Share precompiled baseline ICs across processes"]
+
+    root --> cacheSelfHostedJitCode --> cacheRegularJitCode
+    root --> precompileBaselineICs -->  shareBaselineICs
+
+    shareBaselineICs --> inBinaryStubs
+
+    sharePrecompiledBaselineAcrossProcesses[/"Share Precompiled Baseline JIT Code Across Processes"/]
```
Comment (iainireland) [inline tech_tree/diagram.mmd:170]: In my head this was about self-hosted code specifically, and it was an intermediate step between Bryan's current work and in-binary self-hosted code.

I'm not sure how to express the dependencies between eg "In-Binary CacheIR Stubs" and "In-Binary Self-hosted Baseline Code". Once we've done one of them and solved the cross-compilation issues, the other is presumably easier, but I don't think there's a strict ordering. I guess "Embed Generated Code in Binary" is where we solve the cross-compilation issues, in which case maybe we should get rid of the edge from "In-Binary CacheIR Stubs" to "In-Binary Self-hosted Baseline Code", and instead add an edge from "Cross-process Self-hosted Baseline Code" to "In-Binary Self-hosted Baseline Code"?

**Review by iainireland — APPROVED** [review]
Looks good modulo nits.

File: tech_tree/diagram.mmd (hunk 1 above)
Comment (mgaudet) [reply tech_tree/diagram.mmd]: Yep. Comment updated.

File: tech_tree/index.md (hunk 2 above)
Comment (mgaudet) [reply tech_tree/index.md]: Fixed!

File: tech_tree/diagram.mmd (hunk 3 above)
Comment (mgaudet) [reply tech_tree/diagram.mmd:170]: Ok, I've tried to address this. (that whole region of the diagram is getting a bit rats-nested unfortunately. May need a sub-graph cleanup)

Comment (iainireland) [reply tech_tree/diagram.mmd:170]: At first glance the arrows look right to me, although I agree that it's getting a little messy. Seems fine to tidy it up later.

**Pull request merged**